From af8bc104fdc335efd94f19a33f4f1345f0cd0046 Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Wed, 25 Jan 2023 17:40:42 -0500
Subject: [PATCH] Work around nasm bug with empty assembly files

If you pass an empty assembly file into nasm, it crashes. Add a dummy
instruction which the static linker will hopefully dropped. (This is a
no-op unless you try to link all the assembly files together for a
simpler build.)

Bug: 542
Change-Id: Idd2b96c129a3a39d5f21e3905762cc34c720f6b2
Reviewed-on: https://boringssl-review.googlesource.com/c/boringssl/+/56326
Commit-Queue: David Benjamin <davidben@google.com>
Reviewed-by: Adam Langley <agl@google.com>
(cherry picked from commit a43c76dbe30d619188dc685b7d432a92e7c2b66b)
---
 crypto/perlasm/x86_64-xlate.pl | 7 ++++++-
 crypto/perlasm/x86asm.pl       | 7 ++++++-
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/crypto/perlasm/x86_64-xlate.pl b/crypto/perlasm/x86_64-xlate.pl
index f9068287a..68f1a695a 100755
--- a/crypto/perlasm/x86_64-xlate.pl
+++ b/crypto/perlasm/x86_64-xlate.pl
@@ -1609,7 +1609,12 @@ if ($masm) {
 #endif
 ___
 } elsif ($nasm) {
-    print "\%endif\n";
+    print <<___;
+\%else
+; Work around https://bugzilla.nasm.us/show_bug.cgi?id=3392738
+ret
+\%endif
+___
 } else {
     die "unknown assembler";
 }
diff --git a/crypto/perlasm/x86asm.pl b/crypto/perlasm/x86asm.pl
index 743c79288..d66255ed6 100644
--- a/crypto/perlasm/x86asm.pl
+++ b/crypto/perlasm/x86asm.pl
@@ -290,7 +290,12 @@ ___
 \%ifidn __OUTPUT_FORMAT__, win32
 ___
         print @out;
-        print "\%endif\n";
+        print <<___ unless $masm;
+\%else
+; Work around https://bugzilla.nasm.us/show_bug.cgi?id=3392738
+ret
+\%endif
+___
     } else {
         my $target;
         if ($elf) {
